### DISCUSSION TITLE
`FallibleTransformer` syntax for Coproducts

### DISCUSSION THREAD

**Discussion by doohochang** [arainko/ducktape#57] [Q&A] [answered]
Is there any plan to support `FallibleTransformer` syntax for coproduct types (i.e `sealed trait`, `enum` or `union`) ?

I'm currently using Ducktape for transformations between generated protobuf objects and domain objects, and it's usually fallible to transform a coproduct type from presentation layer to pure domain layer.

Supporting syntax such as `Case.fallibleConst` `Case.fallibleComputed` as same as the current style would be great.

**Comment by arainko** [comment]
Coproduct transformation need more love in general - I can imagine a mechanism that derives total or fallible transformations for coproduct elements that are products and whose names match just like `ducktape` does for your run of the mill case classes.

Can you provide me with an example of how your fallible coproduct transfomations look like right now? I'd be great to get some in-the-field examples to determine a path to automation for this.

**Reply by doohochang** [reply]
Thanks for your consideration!
My cases don't require any explicit validation, but generally include unspecified / unrecognized values for coproduct transformation as below.
```scala
// Coproduct sample in domain layer
sealed trait Animal
object Animal:
  case object Cat extends Animal
  case object Dog extends Animal

// Simplified protobuf enum coproduct sample in presentation layer
sealed trait AnimalProtobuf
object AnimalProtobuf:
  case object Cat extends AnimalProtobuf
  case object Dog extends AnimalProtobuf
  case object Unspecified extends AnimalProtobuf
  case object Unrecognized extends AnimalProtobuf

// Then fallible transformation would be...
def transform(source: AnimalProtobuf): Either[Failure, Animal] =
  import AnimalProtobuf.*
  source match
    case Cat => Right(Animal.Cat)
    case Dog => Right(Animal.Dog)
    case Unrecognized => Left(Failure.UnrecognizedEnum)
    case Unspecified => Left(Failure.UnspecifiedEnum)
```

P.S. We currently use `zio.prelude.Validation` as a data type and support multiple failure cases for transformation via `NonEmptyChunk`.

**Reply by arainko** [reply, marked as answer]
This should be possible now that #71 is in :smile:
